### PR TITLE
[Segmentation] Enable feature vector output for MPA Segmentation

### PR DESCRIPTION
**Pull request by eugene123tw**
Corresponding PRs: 
* MPA submodule: https://github.com/openvinotoolkit/model_preparation_algorithm/pull/26
* mmseg submodule: https://github.com/openvinotoolkit/mmsegmentation/pull/92

### CODE REVIEW

**Comment by eugene123tw** [comment]
run ote-test

File: external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py [hunk 1]
```diff
@@ -201,45 +202,48 @@ def _init_test_data_cfg(self, dataset: DatasetEntity):
...
-
-                result_media = ResultMediaEntity(name=f'{label.name}',
-                                                 type='Soft Prediction',
-                                                 label=label,
-                                                 annotation_scene=dataset_item.annotation_scene,
-                                                 roi=dataset_item.roi,
-                                                 numpy=result_media_numpy)
-                dataset_item.append_metadata_item(result_media, model=self._task_environment.model)
+    def _add_predictions_to_dataset(self, prediction_results, dataset, save_mask_visualization):
+        """ Loop over dataset again to assign predictions. Convert from MMDetection format to OTE format. """
```
Comment (JihwanEom) [inline external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py]: Minor suggestion, could you revise from "MMDetection" to "MMSegmentation" or "MMSeg"?

Comment (eugene123tw) [reply external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py]: Fixed. Thank you

**Review by goodsong81 — CHANGES_REQUESTED** [review]
Inference task looks good to me. Could you add OpenVINO IR support for feature vector / map output?

**Comment by eugene123tw** [comment]
run ote-test

**Comment by eugene123tw** [comment]
run ote-test

**Comment by JihwanEom** [comment]
Hi @chuneuny-emily , could you review this PR for segmentation task? mmseg submodule is updated on this PR, so I'm worried for cls_incr_segmentor.

File: external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py [hunk 2]
```diff
@@ -208,45 +209,33 @@ def _init_test_data_cfg(self, dataset: DatasetEntity):
...
-                                                 label=label,
-                                                 annotation_scene=dataset_item.annotation_scene,
-                                                 roi=dataset_item.roi,
-                                                 numpy=result_media_numpy)
-                dataset_item.append_metadata_item(result_media, model=self._task_environment.model)
+    def _add_predictions_to_dataset(self, prediction_results, dataset, dump_saliency_map):
+        """ Loop over dataset again to assign predictions. Convert from MMSegmentation format to OTE format. """
+        
+        for dataset_item, (prediction, feature_vector, saliency_map) in zip(dataset, prediction_results):
+            soft_prediction = np.transpose(prediction[0], axes=(1, 2, 0))
```
Comment (chuneuny-emily) [inline external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py:216]: In line 216, "ValueError: axes don't match array" error occurs while training.

Can you verify with this command? 
ote train ./external/model-preparation-algorithm/configs/segmentation/ocr-lite-hrnet-18-mod2/template.yaml \
--train-ann-files data/segmentation/custom/annotations/training \
--train-data-roots data/segmentation/custom/images/training \
--val-ann-files data/segmentation/custom/annotations/training \
--val-data-roots data/segmentation/custom/images/training \
--save-model-to data/seg-test \
params --learning_parameters.num_iters 2

**Review by chuneuny-emily — CHANGES_REQUESTED** [review]
Side effect occurs while training. Please check my comment.

File: external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py (hunk 2 above)
Comment (eugene123tw) [reply external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py:216]: @chuneuny-emily The command works for me. Can you check that you have updated the submodule with git submodule update?
```
2022-07-28 11:05:45,385 | INFO : +-------+-------+-------+
|  aAcc | mDice |  mAcc |
+-------+-------+-------+
| 41.71 | 31.62 | 51.06 |
+-------+-------+-------+
2022-07-28 11:05:45,490 | INFO : Now best checkpoint is saved as best_mDice_epoch_2.pth.
2022-07-28 11:05:45,490 | INFO : Best mDice is 0.3162 at 2 epoch.
2022-07-28 11:05:45,490 | INFO : Exp name: /tmp/MPA-task-14nfowop/stage00_SegTrainer-train
2022-07-28 11:05:45,490 | INFO : Epoch(val) [2][4]	current_iters: 2, aAcc: 0.4171, mDice: 0.3162, mAcc: 0.5106, Dice.background: 0.0535, Dice.target: 0.5788, Acc.background: 0.0276, Acc.target: 0.9935
2022-07-28 11:05:46,493 | INFO : run task done.
2022-07-28 11:05:46,493 | INFO : Final model performance: Performance(score: 0.31620000000000004, dashboard: (11 metric groups))
2022-07-28 11:05:46,494 | INFO : called save_model: /tmp/MPA-task-14nfowop/stage00_SegTrainer-train/best_mDice_epoch_2.pth
2022-07-28 11:05:46,656 | INFO : train done.
2022-07-28 11:05:46,659 | INFO : infer()
2022-07-28 11:05:46,659 | INFO : initializing....
2022-07-28 11:05:46,659 | INFO : called _init_recipe()
2022-07-28 11:05:46,659 | INFO : train type = Incremental
2022-07-28 11:05:46,677 | INFO : initialized recipe = /home/yuchunli/git/training_extensions/external/model-preparation-algorithm/submodule/recipes/stages/segmentation/class_incr.py
2022-07-28 11:05:46,680 | INFO : initialized.
2022-07-28 11:05:46,680 | INFO : running task... kwargs = {'dump_features': True, 'dump_saliency_map': False}
2022-07-28 11:05:46,680 | INFO : called build_recipe()
2022-07-28 11:05:46,680 | INFO : seems to be passed stage yaml...
2022-07-28 11:05:46,680 | INFO : build_stage()
2022-07-28 11:05:46,681 | INFO : Training seed was set to 42 w/ deterministic=False.
2022-07-28 11:05:46,681 | INFO : work dir = /tmp/MPA-task-14nfowop/stage00_SegInferrer-train
2022-07-28 11:05:46,681 | INFO : CUDA_VISIBLE_DEVICES = None
2022-07-28 11:05:46,681 | INFO : configured logger at /tmp/MPA-task-14nfowop/stage00_SegInferrer-train with named 20220728_110546.log
2022-07-28 11:05:46,681 | INFO : configure!: training=False
Use load_from_local loader
2022-07-28 11:05:46,734 - mmseg - INFO - task config!!!!: training=False
Use load_from_local loader
2022-07-28 11:05:46,786 - mmseg - INFO - infer!
2022-07-28 11:05:46,787 - mmseg - INFO - Inferring on input source: data.test
2022-07-28 11:05:46,787 | INFO : test_mode : False
Use load_from_local loader
2022-07-28 11:05:46,922 - mmseg - INFO - ----------------- ClassIncrSegmentor.load_state_dict_pre_hook() called w/ prefix: 
2022-07-28 11:05:46,922 - mmseg - INFO - ['background', 'target'] -> ['background', 'target'] ([0, 1])
2022-07-28 11:05:47,276 | INFO : run task done.
2022-07-28 11:05:47,343 | INFO : called evaluate()
2022-07-28 11:05:47,343 | INFO : Computing mDice
function [__compute_dice_averaged_over_pixels] finished in 49.02982711791992 ms
2022-07-28 11:05:47,392 | INFO : mDice after evaluation: 0.5799746697654503
Performance(score: 0.5799746697654503, dashboard: (1 metric groups))
```

Comment (chuneuny-emily) [reply external/model-preparation-algorithm/mpa_tasks/apis/segmentation/task.py:216]: I used eugene/CVS-86547-enable-saliency-mpa-seg for both ote and submodule.

**Review by chuneuny-emily — DISMISSED** [review]

**Review by chuneuny-emily — CHANGES_REQUESTED** [review]
Please fix the side effect on mmsegmentation first before merging this PR, and update mmseg submodule commit.

**Review by chuneuny-emily — DISMISSED** [review]
No side effect on train/eval/export/optimize.

**Review by JihwanEom — DISMISSED** [review]
LGTM

**Review by JihwanEom — APPROVED** [review]
LGTM, could you add descriptions on this PR about OTE_SDK version update for deployment code or CVS-84981 (OMZ) ?

File: external/model-preparation-algorithm/tests/ote_cli/test_seg_cls_il.py [hunk 3]
```diff
@@ -106,7 +106,6 @@ def test_ote_demo_openvino(self, template):
     @e2e_pytest_component
     @pytest.mark.parametrize("template", templates, ids=templates_ids)
     def test_ote_deploy_openvino(self, template):
-        pytest.xfail("Known issue CVS-84981")
```
Comment (harimkang) [inline external/model-preparation-algorithm/tests/ote_cli/test_seg_cls_il.py:109]: What does this have to do with this PR? Is it related to OTE SDK update?

Comment (eugene123tw) [reply external/model-preparation-algorithm/tests/ote_cli/test_seg_cls_il.py:109]: yes, the segmentation deployment code was not working due to the old OTE SDK doesn't have `check_input_parameters_type`. It's fixed by upgrading OTE SDK.

**Review by chuneuny-emily — APPROVED** [review]

**Review by harimkang — APPROVED** [review]

**Review by goodsong81 — APPROVED** [review]
